### DISCUSSION TITLE
Debugging incorrect compiler usage

### DISCUSSION THREAD

**Discussion by yawhide** [sqlalchemy/sqlalchemy#7453] [Usage Questions]
Hello,

I have an odd problem that I am having trouble figuring out. I have a particular test case that is using the SQLite compiler and sqlite database instead of using the Postgresql compiler and postgresql database and I cannot figure out how this is happening. I am using Flask-SQLalchemy so let me know if I should be asking them instead of y'all.

Things I have tried:
1. log the SQLALCHEMY_DATABASE_URL when I init the app and right before I run the problematic sql (both times the url is `postgresql://...`)
1. Specifically use an `insert` using the postgresql dialect (I get an error saying `sqlalchemy.exc.CompileError: RETURNING is not supported by this dialect's statement compiler.` because I do `insert(User).values({"email": faker.email()}).returning(User.id)`. I guess sqlite doesnt have returning.

I am at a lost. I dont understand why sqlite is being used :(
This doesn't happen locally, only on my CI so I have been trying to do a lot of print debugging to no avail.

Any ideas how to figure out what is going on and debug this issue? 

Thanks

**Comment by zzzeek** [comment]
I would see if you are using "bound metadata" anywhere, like MetaData(bind=engine).  Get rid of any of that for starters.

other than that it depends on how this SQL is being generated.  is this like, connection.execute()?  or are you calling .compile() on a SQL construct?

**Reply by yawhide** [reply]
Thanks for the reply. We only use `MetaData(naming_conversions)`. 

We use the v1.3 ORM api. Stuff like 
```
user = User(email='blah)
db.session.add(user)
db.session.commit()
```

or 
```
users = db.session.query(User).filter(...)
```

Is there some introspection I can do on db.session to check why it is generating sqlite sql vs postgresql sql?

**Reply by zzzeek** [reply]
well grepping your source code for "sqlite" as well as accounting for everywhere you call create_engine() would be a start.  another way, something I do often though it's not everyone's style, I just go right into the library in question (like, I vi packages inside of /usr/lib/python or whatever) and just stick pdbs or assertions right in there.  like if you went into the SQLite dialect constructor in sqlalchemy/dialects/sqlite/base.py and just put a big "assert False" in there, you'd get an instant stack trace where that dialect is being built up.
